### ISSUE TITLE
"auto_create" not working, can't login using LDAP unless account manually created in MySQL first.

### ISSUE THREAD

**Issue by echoxxzz** [ampache/ampache#2865]
Trying to setup Ampache to authenticate against Samba AD server but it doesn't work and nothing ever shows up in the logs showing Ampache trying make an LDAP connection unless the account already exists in MySQL (then I see lots of LDAP info in the logs).

Once I setup an account in MySQL with the same username that I'm trying to authenticate with against LDAP (but with a different password than the one used for LDAP) I can login using the LDAP. I can even logout and login again using the same account but with the other password store in MySQL.

My ampache.cfg.php files contains these lines:

auth_methods = "ldap,mysql"
auto_create = "true"
auto_user = "user"
ldap_url = "ldap://dc.mydomain.com"
ldap_username = "cn=ampauth,cn=Users,dc=mydomain,dc=com"
ldap_password = "PASSWORD"
ldap_search_dn = "cn=Users,dc=mydomain,dc=com"
ldap_objectclass = "user"
ldap_filter = "(uid=%v)"
ldap_name_field = "displayName"
ldap_email_field = "mail"
ldap_start_tls = "true"

**Comment by echoxxzz** [comment]
Went back to version 4.4.0 and it's working as expected. This bug seems to only effect version 4.4.1.

**Comment by lachlan-00** [comment]
There is a change in auth for tokens but not user pass.

I need to set up a LDAP for this as there seem to be lots of general issues with ldap

**Comment by lachlan-00** [comment]
I think I know what it could be based on the token change... Are there any logs for a no created user so I can follow this up later?

**Comment by gramakri** [comment]
I can reproduce with the Cloudron package (which uses LDAP). LDAP works with 4.4.0 but not with 4.4.1. @lachlan-00 what logs do you need ? Let me see if I can quickly figure what the problem is.

**Comment by gramakri** [comment]
@lachlan-00 The issue is because of 4aa7a30f88e56edfb396234d2ffe7d7758b0dbed . It checks for the username in `get_valid_usernames` [here](https://github.com/ampache/ampache/commit/4aa7a30f88e56edfb396234d2ffe7d7758b0dbed#diff-797a538c67a9dfca25e8f3519c677c957ec885eb33d51ac25e712191f5c40ddaR100) . The username doesn't exist in DB since it is not auto-created yet.

**Comment by lachlan-00** [comment]
16fa60950 gets that now.

allow it to fall through to ldap instead of checking right at login attempt

**Comment by gramakri** [comment]
@lachlan-00 thanks! I guess this issue can be closed. Will this be in 4.4.2?

**Comment by lachlan-00** [comment]
That's right. Probably a week or 2. There are a couple of goods fixes that need to go out

**Comment by NatureMuncher** [comment]
Here is a workaround for others who want to add multiple ldap users.

Run the following from a Windows PowerShell console:
~~~
Get-ADUser -filter {Enabled -eq $True} | Select-Object -Property "sAMAccountName"
~~~
which dumps a list of AD login names. 

Remove system accounts, then paste the good account names into a txt file on your Ampache server under `/var/www/html/ampache/bin` called `users.txt` and run the following as the apache user (from the bin directory):
~~~
while read -r LINE; do ./cli admin:addUser $LINE -e $LINE@my-email-domain.com -l 75 ; done < users.txt
~~~
The `75` makes them catalog managers. It auto-generates a random local user's password when creating the user, then ensure your `ampache.cfg.php` file checks the LDAP settings before the SQL entries for authentication with:
`auth_methods = "ldap,mysql"`

**Comment by lachlan-00** [comment]
> I see this is still open. Here are the logs for my system when a user attempts to log in and does not have an account:
> 
> ```
> 2022-03-09T16:01:38+00:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
> 2022-03-09T16:01:38+00:00 [ampache] (Ampache\Module\System\Dba) -> Database connection...
> 2022-03-09T16:01:38+00:00 [] (Ampache\Module\Application\ApplicationRunner) -> Found handler "Ampache\Module\Application\Login\DefaultAction" for action "default"
> 2022-03-09T16:01:38+00:00 [] (Ampache\Module\System\Session) -> Unable to read value from key [...] no data found
> 2022-03-09T16:01:38+00:00 [] (Ampache\Module\Authentication\Ldap\Ldap) -> binding with username `cn=server_user_search,OU=service-accounts,OU=sorted - users,DC=sub,DC=domain,DC=com`
> 2022-03-09T16:01:38+00:00 [] (Ampache\Module\Authentication\Ldap\Ldap) -> search: (&(objectclass=organizationalPerson)(sAMAccountName=my_user))
> 2022-03-09T16:01:38+00:00 [] (Ampache\Module\Authentication\Ldap\Ldap) -> searching in `ou=people,ou=sorted,dc=sub,dc=domain,dc=com` for `(&(objectclass=organizationalPerson)(sAMAccountName=my_user))`
> 2022-03-09T16:01:38+00:00 [] (Ampache\Module\Authentication\Ldap\Ldap) -> binding with username `CN=My User,OU=folks,OU=sorted - users,DC=int,DC=domain,DC=com`
> 2022-03-09T16:01:38+00:00 [] (Ampache\Module\Authentication\Ldap\Ldap) -> Return value of authentication: {"success":true,"type":"ldap","username":"my_user","name":"My User","email":"my_user@domain.com"}
> ```
> 
> Creating the account with a nonsense password allows them to log in with their ldap password.

so for you the ldap module returns true but the account is still missing?

going to be looking at this now

**Comment by NatureMuncher** [comment]
Yes, the above was returned for a valid AD user who did not yet have an account in Ampache, and no account was created in Ampache.

**Comment by lachlan-00** [comment]
Thanks. I'll set up an ldap server again and fix it all up next week

**Comment by lachlan-00** [comment]
can anyone provide some logs on this one right now?

I'm going to try and close this off before we go to 6.0.0

**Comment by gramakri** [comment]
@lachlan-00 Atleast on Cloudron LDAP, this works fine. Users are auto_created on first login

**Comment by lachlan-00** [comment]
had a check and the auth happens before the check now and is probably fine now for this specific issue
